jsregexp.lua

```diff
@@ -0,0 +1,164 @@
+    local result = {}
+
+    local i = 1
+    local repl_len = #replacement
+
+    while true do
+        local j = string.find(replacement, "$", i, true)
+        if j == nil or j + 1 > repl_len then break end
+        table.insert(result, string.sub(replacement, i, j - 1))
+        local j0 = j
+        local c = string.sub(replacement, j + 1, j + 1)
+        j = j + 2
+        if c == '$' then
+            table.insert(result, "$")
+        elseif c == '&' then
+            table.insert(result, match[0])
+        elseif c == '`' then
+            table.insert(result, string.sub(str, 1, match.index))
+        elseif c == '\'' then
+            table.insert(result, string.sub(str, match.index + #match[0]))
+        elseif is_digit(c, 1) then
+            local k = c
+            local kv
+            local dig2 = false
+            if j <= repl_len and is_digit(replacement, j) then
+                k = k .. string.sub(replacement, j, j)
+                dig2 = true
+            end
+            local kv1 = tonumber(k)
+            assert(kv1 ~= nil)
+
+            -- This behavior is specified in ES6 and refined in ECMA 2019 
+            if dig2 and kv1 >= 1 and match[kv1] ~= nil then
+                kv = kv1
+                j = j + 1
+            else
+                kv = tonumber(k)
+                assert(kv ~= nil)
+            end
+            if kv >= 1 and match[kv] ~= nil then
+                table.insert(result, match[kv])
+            else
+                table.insert(result, string.sub(replacement, j0, j))
+            end
+        elseif c == '<' and match.groups ~= nil then
+            local k = string.find(replacement, ">", j, true)
+            if k == nil then
+                table.insert(result, string.sub(replacement, j0, j))
+            else
+                local name = string.sub(replacement, j, k - 1)
+                local capture = match.groups[name]
+                assert(capture ~= nil, "invalid capture name: " .. name)
+                table.insert(result, capture)
+                j = k + 1
+            end
+        else
+            table.insert(result, string.sub(replacement, j0, j))
+        end
+
+        i = j
+    end
+    table.insert(result, string.sub(replacement, i))
+    return table.concat(result)
+end
+
+function jsregexp.mt.replace(re, str, replacement)
+
+    local jstr = jsregexp.to_jsstring(str)
+
+    re.last_index = 1
+
+    local output = {}
+
+    local prev_index = 1
+    local cur_index = 1
+    while true do
+        prev_index = re.last_index
+        local match = re:exec(jstr)
+        if match == nil then break end
+        cur_index = re.last_index
+
+        table.insert(output, string.sub(str, prev_index, match.index - 1))
+        if type(replacement) == "function" then
+            table.insert(output, replacement(match, str))
+        else
+            table.insert(output, get_substitution(match, str, replacement))
+        end
```

Comment:
Being able to pass a table as a replacement (like in `gsub`) would be nice.

Reply:
Sounds good to add in a subsequent PR. Want to keep the scope of this PR down.